lib/ansible/modules/packaging/os/yum.py

```diff
@@ -742,46 +802,43 @@ def install(module, items, repoq, yum_basecmd, conf_file, en_repos, dis_repos, i
+            for package in pkglist:
+                # Get the NEVRA of the requested package using pkglist instead of spec because pkglist
+                #  contains consistently-formatted package names returned by yum, rather than user input
+                #  that is often not parsed correctly by splitFilename().
+                (name, ver, rel, epoch, arch) = splitFilename(package)
+
+                # Check if any version of the requested package is installed
+                inst_pkgs = is_installed(module, repoq, name, conf_file, en_repos=en_repos, dis_repos=dis_repos, is_pkg=True)
+                if inst_pkgs:
+                    (cur_name, cur_ver, cur_rel, cur_epoch, cur_arch) = splitFilename(inst_pkgs[0])
+                    compare = compareEVR((cur_epoch, cur_ver, cur_rel), (epoch, ver, rel))
+                    if compare > 0:
+                        downgrade_candidate = True
+                    else:
+                        downgrade_candidate = False
+                        break
+
+            # If package needs to be installed/upgraded/downgraded, then pass in the spec
             # we could get here if nothing provides it but that's not
             # the error we're catching here
             pkg = spec
 
         pkgs.append(pkg)
 
-    if pkgs:
-        cmd = yum_basecmd + ['install'] + pkgs
-
-        if module.check_mode:
-            module.exit_json(changed=True, results=res['results'], changes=dict(installed=pkgs))
-
-
-        lang_env = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C')
-        rc, out, err = module.run_command(cmd, environ_update=lang_env)
-
-        if rc == 1:
-            for spec in items:
-                # Fail on invalid urls:
-                if '://' in spec and ('No package %s available.' % spec in out or 'Cannot open: %s. Skipping.' % spec in err):
-                    module.fail_json(msg='Package at %s could not be installed' % spec, rc=1, changed=False)
-
-        res['rc'] = rc
-        res['results'].append(out)
-        res['msg'] += err
-        res['changed'] = True
-
-        # special case for groups
-        if spec.startswith('@'):
-            if ('Nothing to do' in out and rc == 0) or ('does not have any packages to install' in err):
-                res['changed'] = False
+        if downgrade_candidate and allow_downgrade:
+            downgrade_pkgs.append(pkg)
+        else:
+            pkgs.append(pkg)
 
-        if rc != 0:
-            res['changed'] = False
-            module.fail_json(**res)
+    if downgrade_pkgs:
+        res = exec_install(module, items, 'downgrade', downgrade_pkgs, res, yum_basecmd)
 
-        # FIXME - if we did an install - go and check the rpmdb to see if it actually installed
-        # look for each pkg in rpmdb
-        # look for each pkg via obsoletes
+    if pkgs:
+        res = exec_install(module, items, 'install', pkgs, res, yum_basecmd)
```

Comment:
So if `downgrade_pkgs` is not empty and `pkgs` is not empty, you rewrite some of the `res` content here via `exec_install` (like `changed`), no?